### PR TITLE
Posts: Cleaner Post Controls component

### PR DESCRIPTION
**Pull request by Copons**
As discussed in #7451 and attempted before-time in #7574, I rewrote the `PostControls` component in ES6 and moved the controls logic out of the render method for improved readability.

_(Please disregard the "Redux-ify" part in the commit message. I'm used to Redux-ify components, but not this time!)_

### CODE REVIEW

File: client/my-sites/posts/post-controls.jsx [hunk 1]
```diff
@@ -1,244 +1,238 @@
...
-			availableControls = [],
-			extraControls = [],
-			parsed, previewURL, statsURL, postControlsClass, mainControls, moreControls, i, moreControlsSpliceIndex;
+import { isEnabled } from 'config';
+import { ga } from 'lib/analytics';
+import { userCan } from 'lib/posts/utils';
+import Gridicon from 'components/gridicon';
+
+export class PostControls extends PureComponent {
+
```
Comment (retrofox) [inline client/my-sites/posts/post-controls.jsx]: Is it needed this break line? I think we aren't used to adding it.

**Comment by retrofox** [comment]
I wonder if we could take advantage of this refactoring and re-implement the component as a PureComponent since it seems to be stateless component, no refs usage, etc.

**Comment by Copons** [comment]
@retrofox It is already a `PureComponent`. Do you mean functional stateless?

**Comment by retrofox** [comment]
> @retrofox It is already a PureComponent. Do you mean functional stateless?

Yup, sorry. You're right.
I meant to `Functional Stateless Component` instead of inherits from the `PureComponent` class.

**Comment by Copons** [comment]
Here you go @retrofox

**Comment by retrofox** [comment]
You rocks!

**Comment by retrofox** [comment]
There is an erratic behavior when the viewport is resized. In some moment (see the animation) the primary state loses some buttons moving them to the secondary section, shown through of `... More` button.
When I do a hard-refresh the buttons are shown again.



it seems to happen in prod though.

File: client/my-sites/posts/post-controls.jsx [hunk 2]
```diff
@@ -1,244 +1,223 @@
...
-				target: null,
-				onClick: this.edit,
-				icon: 'pencil'
+import { isEnabled } from 'config';
+import { ga } from 'lib/analytics';
+import { userCan } from 'lib/posts/utils';
+import Gridicon from 'components/gridicon';
+
+const view = () => ga.recordEvent( 'Posts', 'Clicked View Post' );
+const preview = () => ga.recordEvent( 'Posts', 'Clicked Preiew Post' );
```
Comment (retrofox) [inline client/my-sites/posts/post-controls.jsx]: typo

**Comment by retrofox** [comment]
Beyond my comments, it looks good to me. @gwwar , do you mind take a quick review here when you have the chance?

File: client/my-sites/posts/post-controls.jsx [hunk 3]
```diff
@@ -1,244 +1,223 @@
...
+const edit = () => ga.recordEvent( 'Posts', 'Clicked Edit Post' );
+const copy = () => ga.recordEvent( 'Posts', 'Clicked Copy Post' );
+const viewStats = () => ga.recordEvent( 'Posts', 'Clicked View Post Stats' );
```
Comment (gwwar) [inline client/my-sites/posts/post-controls.jsx]: Maybe some better naming here. Eg `recordViewPost`.

Just fyi that we do have a preference for using redux analytics actions instead of using the analytics lib directly. This helps for any future component testing, since analytics assumes window is present.

We can leave connecting the component as a follow up.
See: https://github.com/Automattic/wp-calypso/tree/master/client/state/analytics

File: client/my-sites/posts/post-controls.jsx [hunk 4]
```diff
@@ -1,244 +1,223 @@
...
+		site,
+		translate,
+	} = props;
+	const controls = { main: [], more: [] };
+
+	// NOTE: Currently Jetpack site posts are not returning `post.capabilities`
+	// and those posts will not have access to post management type controls
+
+	// Main Controls (not behind ... more link)
+	if ( userCan( 'edit_post', post ) ) {
```
Comment (gwwar) [inline client/my-sites/posts/post-controls.jsx]: Also note that if this component is connected in the future, we can use the following in mapStateToProps
```
import { canCurrentUser } from 'state/current-user/selectors';
//...
const userCanEditPost = canCurrentUser( state, siteId, 'edit_post');
```

File: client/my-sites/posts/post-controls.jsx [hunk 5]
```diff
@@ -1,244 +1,223 @@
...
+			icon: 'chevron-left',
+			onClick: onHideMore,
+			text: translate( 'Back' ),
+		} );
+	}
+
+	return controls;
+};
+
+const getControlElements = controls => controls.map( ( control, index ) =>
```
Comment (gwwar) [inline client/my-sites/posts/post-controls.jsx]: If we want this to exist as it's own component, please split this out into it's own file. Otherwise I'd recommend `extending Component` instead with helper render methods

**Comment by gwwar** [comment]
Anyway, feel free to :shipit: if @retrofox has tested for regressions.

Please keep in mind to try and use stateless functional components for short/simple components, and to keep one component per file. Things get rather hard to follow otherwise.

File: client/my-sites/posts/post-controls.jsx (hunk 5 above)
Comment (Copons) [inline client/my-sites/posts/post-controls.jsx]: Just to be clear: you're suggesting of moving the single controls into their own components (maybe creating a folder like `my-sites/posts/post-controls/` containing this one I'm refactoring and a new `control.jsx`), and move the `map` into the render.
This seems a good idea and would also reorder a little the `my-sites/posts/` folder.

**Comment by Copons** [comment]
If you're ok with it, I'd like to keep on working on this, connecting the component and moving the control elements into their own file. It should not take much time and new LOCs.
If instead you'd rather have smaller PRs to review, I can ship this and move on with a new one.

File: client/my-sites/posts/post-controls.jsx (hunk 5 above)
Comment (gwwar) [inline client/my-sites/posts/post-controls.jsx]: That right.

**Comment by gwwar** [comment]
> If you're ok with it, I'd like to keep on working on this, connecting the component and moving the control elements into their own file.

All the things I mentioned can be done as follow up PRs 👍  Thanks for taking the time to tidy this.

**Comment by Copons** [comment]
@gwwar Great! I'll tackle them as soon as I ship this, then.

**Comment by Copons** [comment]
Rebased to fix conflicts introduced with https://github.com/Automattic/wp-calypso/pull/10772.
I've tested the new Publicize logic and it work on my JP sites.
Looping in @gziolo for a quick check if I've missed anything.

**Comment by gziolo** [comment]
Looks good 👍

**Comment by retrofox** [comment]
LGTM as well.

File: client/my-sites/posts/post-controls.jsx [hunk 6]
```diff
@@ -1,251 +1,228 @@
...
+export const PostControls = props => {
+	const { main, more } = getAvailableControls( props );
+	const classes = classNames(
+		'post-controls',
+		{ 'post-controls--desk-nomore': more <= 2 }
+	);
+
+	return (
+		<div className={ classes }>
+			{ more.length &&
```
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: This condition is causing `0` to be output when there are no "More" items. You can see this on the All Sites Drafts screen:

image

https://wordpress.com/posts/my/drafts

The way the condition reads, since the first part is not truthy, it uses the value on the left hand of the `&&`, and React will happily output `0`. You need this to either evaluate to a boolean, `null`, or `undefined`.

I suggest changing to `more.length > 0 &&`

Comment (Copons) [inline client/my-sites/posts/post-controls.jsx]: Dang, just deployed. Coming up with a quick fix
